File: examples/ospf-kotlin-example/src/main/fuookami/ospf/kotlin/example/framework_demo/demo1/route_context/Aggregation.kt
package fuookami.ospf.kotlin.example.framework_demo.demo1.route_context

import fuookami.ospf.kotlin.utils.functional.*
import fuookami.ospf.kotlin.core.frontend.model.mechanism.*
import fuookami.ospf.kotlin.example.framework_demo.demo1.route_context.model.Assignment
import fuookami.ospf.kotlin.example.framework_demo.demo1.route_context.model.Graph
import fuookami.ospf.kotlin.example.framework_demo.demo1.route_context.model.Service

class Aggregation(
    val graph: Graph,
    val services: List<Service>,
    val assignment: Assignment
) {
    fun register(model: LinearMetaModel): Try {
        val subprocesses = arrayListOf(
            { return@arrayListOf assignment.register(model) }
        )

        for (subprocess in subprocesses) {
            when (val result = subprocess()) {
                is Failed -> {
                    return Failed(result.error)
                }

                is Ok -> {}
            }
        }
        return ok
    }
}

File: ospf-rust-example/src/framework/demo1/route_context/aggregation.rs
use super::model::{Assignment, Edge, Graph, Node, Service};

pub struct Aggregation {
    pub graph: Graph,
    pub services: Vec<Service>,
    pub assignment: Assignment,
}

impl Aggregation {
    pub fn new(graph: Graph, services: Vec<Service>, assignment: Assignment) -> Self {
        Self {
            graph,
            services,
            assignment,
        }
    }
}
